File: 0050-powx-n/0050-powx-n.java
class Solution {
    public double myPow(double x, int n) {
        double finalAnswer = 1.0; // Variable to store the final result
        long powerOf = n; // Use a long to handle the edge case of n being Integer.MIN_VALUE

        // If the power is negative, convert it to positive for easier calculation
        if (powerOf < 0) {
            powerOf = -1 * powerOf;
        }

        // Loop until all bits of the power have been processed
        while (powerOf > 0) {
            // If the current power is even
            if (powerOf % 2 == 0) {
                x = x * x; // Square the base
                powerOf = powerOf / 2; // Halve the power
            } else { // If the current power is odd
                finalAnswer = finalAnswer * x; // Multiply the result by the base
                powerOf = powerOf - 1; // Decrement the power by 1
            }
        }

        // If the original power was negative, invert the result
        if (n < 0) {
            finalAnswer = 1.0 / finalAnswer;
        }
        return finalAnswer; // Return the final result
    }
}

File: 0050-powx-n/0050-powx-n.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        # Recursive helper function to handle the power calculation
        def helper(x, n):
            # Base case: Any number raised to the power of 0 is 1
            if n == 0:
                return 1
            # If n is even, recursively call the function with n/2 and square the result
            elif n % 2 == 0:
                ans = helper(x, n // 2)
                return ans * ans
            # If n is odd, multiply x with the result of the function called with (n-1)
            else:
                return x * helper(x, n - 1)

        # Handle the case where n is negative
        if n < 0:
            x = 1 / x
            n = -n
        
        # Call the helper function with the original x and n values
        return helper(x, n)
